Gap automeshing

dbochkov-flexcompute opened the pull request:
Integrated gap meshing in an iterative way. Now `LayerRefinementSpec` has two additional parameters:
- `gap_meshing_iters: NonNegativeInt = 1` to set the number of iteration to perform for attempting resolving all gaps
- `dl_min_from_gap_width: bool = True` whether or not to reduce `dl_min` reduce to the minimal detected gap width.

A few of points to discuss:
- Doing just one iterations seems to be performing quite well, so that could be a good default. But maybe we should switch to a higher number and just let it iterate until no multiple crossings are detected in a single cell?
- Currently it does not distinguish between thin gaps and thin strips. Do we want to control mesher behavior for the two types
- while we use snapping planes for achieving the desired result, we currently do not store them, and, as a consequence, we cannot plot them. Should we worry about that?

examples of meshing thin strips and gaps:
<img width="1350" alt="Screenshot_20250417_231504" src="https://github.com/user-attachments/assets/77f485ec-dea6-4892-8015-f113f1dd758c" />
<img width="1347" alt="Screenshot_20250417_231652" src="https://github.com/user-attachments/assets/237f7efb-7abb-4fae-a84b-e04c0d4f41bd" />

dbochkov-flexcompute commented:
one more point:
- currently we also take into account all structures. Should we worry about making it work for a selected set of structures in this PR?

weiliangjin2021 commented:
Thanks for this great feature!

> Doing just one iterations seems to be performing quite well, so that could be a good default. But maybe we should switch to a higher number and just let it iterate until no multiple crossings are detected in a single cell?

Agree that we can set 1 as default if ti works quite well already. Then we can add an option like `inf` or a string to indicate it will iterate until no crossings are detected?

> Currently it does not distinguish between thin gaps and thin strips. Do we want to control mesher behavior for the two types

Right now, we don't have good conformal scheme for thin strips, so probably no need right now. 

> while we use snapping planes for achieving the desired result, we currently do not store them, and, as a consequence, we cannot plot them. Should we worry about that?

Not too concerned, but I wonder that it's not hard to store them?

> currently we also take into account all structures. Should we worry about making it work for a selected set of structures in this PR?

My understanding is that metallic structures are more of a concern. So we probably only need to distinguish between metallic (PEC, lossyMetal) and dielectric.

dmarek-flex commented on tidy3d/components/grid/grid_spec.py:
```diff
@@ -1921,6 +2114,102 @@ def make_grid(
+            If `None`, recomputes internal override structures.
+        internal_snapping_points : List[CoordinateOptional]
+            If `None`, recomputes internal snapping points.
+
```
missing some of the parameters in the docstring

dmarek-flex replied:
looks like missing parameters in the original function's docstring as well

dmarek-flex replied:
`periodic` and `min_dl_from_gaps`

dbochkov-flexcompute replied:
done

weiliangjin2021 commented on tidy3d/components/grid/grid_spec.py:
```diff
@@ -1024,6 +1024,19 @@ class LayerRefinementSpec(Box):
+        "The underlying algorithm detects gaps contained in a single cell and places a snapping plane at the gaps's centers.",
+    )
+
+    dl_min_from_gap_width: bool = pd.Field(
```
Now that we have dl_min from quite a few places: gap refinement, corner refinement, thickness refinement, and minimal feature, shall we allow controling them individually and combine them into a new class?

weiliangjin2021 replied:
(not to be considered in this PR)

weiliangjin2021 commented on tidy3d/components/grid/grid_spec.py:
```diff
@@ -1497,6 +1534,162 @@ def _override_structures_along_axis(
+        cells_ij = []
+        cells_dy = []
+        h_inds = np.argmax(x[:, None] > vertices[None, :, 0], axis=0)
+        h_inds[vertices[:, 0] > x[-1]] = len(x)
```
More informative variable name for `h_inds`

weiliangjin2021 commented on tidy3d/components/grid/grid_spec.py:
```diff
@@ -1497,6 +1534,162 @@ def _override_structures_along_axis(
+        cells_dy = []
+        h_inds = np.argmax(x[:, None] > vertices[None, :, 0], axis=0)
+        h_inds[vertices[:, 0] > x[-1]] = len(x)
+        for ind_beg, ind_end, v_beg, v_end in zip(
```
add a comment that `beg` and `end` refers to the two vertices on the edge of the polygon

weiliangjin2021 commented on tidy3d/components/grid/grid_spec.py:
```diff
@@ -1497,6 +1534,162 @@ def _override_structures_along_axis(
+            if ind_beg > ind_end:
+                ind_beg, ind_end, v_beg, v_end = ind_end, ind_beg, v_end, v_beg
+
+            if ind_end > ind_beg:
```
clearer with `if ind_end == ind_beg: continue`, and undent the rest? This brings up the question on how the close-to-vertical polygon edge is handled. I guess they are handled when looking at horizontal intersection?

weiliangjin2021 commented on tidy3d/components/grid/grid_spec.py:
```diff
@@ -1497,6 +1534,162 @@ def _override_structures_along_axis(
+        """Detect intersection points of single polygon and grid lines."""
+        v_cells_ij, v_cells_dy = self._find_vertical_intersections(x, y, vertices)
+        # reuse the same command but flip dimensions
+        h_cells_ij, h_cells_dx = self._find_vertical_intersections(y, x, vertices[:, [1, 0]])
```
use np.flip?

weiliangjin2021 commented on tidy3d/components/grid/grid_spec.py:
```diff
@@ -1497,6 +1534,162 @@ def _override_structures_along_axis(
+        # reuse the same command but flip dimensions
+        h_cells_ij, h_cells_dx = self._find_vertical_intersections(y, x, vertices[:, [1, 0]])
+        if len(h_cells_ij) > 0:
+            h_cells_ij = np.roll(h_cells_ij, axis=1, shift=1)
```
more explanation here?

weiliangjin2021 commented on tidy3d/components/grid/grid_spec.py:
```diff
@@ -1921,6 +2114,102 @@ def make_grid(
+                            min_gap_width = min(min_gap_width, gap_width)
+
+                if len(new_snapping_lines) == 0:
+                    log.info("Grid is no longer changing. Stopping recursive refinement.")
```
include the information on the number of iteration where it stops.

weiliangjin2021 reviewed (commented):
Half way through the code. Looks quite nice, although taking quite a bit efforts in understanding the logic. Some high-level description in each function will be very helpful. Some minor comments so far:

dbochkov-flexcompute commented:
sprinkled around some more comments in the code, had to recall myself what I was doing :sweat_smile:

dmarek-flex commented on tidy3d/components/grid/grid_spec.py:
```diff
@@ -1497,6 +1534,253 @@ def _override_structures_along_axis(
+
+        # loop over all shapes
+        for mat, shapes in merged_geos:
+            if not mat.is_pec:
```
We should support `LossyMetalMedium` as well, I need to make the same check in my next PR. Maybe we can add `is_metal_conductor` to `AbstractMedium`

dbochkov-flexcompute replied:
actually we rely here on that the function that constructed `merged_geos` also converted all `LossyMetalMedium`'s into pec https://github.com/flexcompute/tidy3d/blob/daniil/gap-meshing/tidy3d/components/grid/corner_finder.py#L109-L116 

Not ideal, as it is very easy not to realize that, but I added a couples of notes to point that out

dmarek-flex commented on tidy3d/components/grid/grid_spec.py:
```diff
@@ -1497,6 +1534,253 @@ def _override_structures_along_axis(
                 override_structures += refinement_structures
         return override_structures
 
+    def _find_vertical_intersections(self, x, y, vertices):
```
add some type hints for the return value of these methods

dbochkov-flexcompute replied:
done

dmarek-flex commented on tidy3d/components/grid/grid_spec.py:
```diff
@@ -1497,6 +1534,253 @@ def _override_structures_along_axis(
+
+        # generate horizontal snapping lines
+        snapping_lines_y = []
+        if len(v_cells_ij) > 0:
```
Seems possible to avoid duplicating the code for the horizontal and vertical lines (as you did for `_find_vertical_intersections`

dbochkov-flexcompute replied:
done

dmarek-flex commented on tidy3d/components/grid/grid_spec.py:
```diff
@@ -1497,6 +1534,253 @@ def _override_structures_along_axis(
+        cells_dy = []
+
+        # for each polygon vertex find the index of the first grid line on the right
+        grid_lines_on_right = np.argmax(x[:, None] > vertices[None, :, 0], axis=0)
```
It might make sense to use a small tolerance, so that we can distinguish between almost exactly coincident with the cell boundary and certainly greater than.

dmarek-flex replied:
That might minimize the number of false positives, could also result in some of the asymmetry present in the resulting grids.

dbochkov-flexcompute replied:
good idea, added special treatment for such cases

dmarek-flex reviewed (commented):
Another great feature! I found a couple of small things in the associated comments. The only real sticking point for me is that it would be nice to get a symmetric grid, when the structure itself is symmetric. 

I think it came up in the meeting and maybe there is no easy solution. But for the structure below it seems to always give an asymmetric grid. I thought maybe part of the reason could be due to how intersections are found with grid boundaries, which does not make use of a tolerance. The grid boundary and the polygon vertex might be very close. 

![image](https://github.com/user-attachments/assets/08205838-5938-468c-a055-43c2f3de41ad)

And here is the original grid without gap refinement. (setting gap_meshing_iters=0)

![Screenshot from 2025-04-23 15-28-32](https://github.com/user-attachments/assets/82daecca-73b4-4d86-8a96-9a8073ece054)

dbochkov-flexcompute commented:
should be all addressed now. Some notable changes:
- Regarding @dmarek-flex point on doing this with some tolerance allowance, I implemented the following: intersections that are very close to cell boundaries are duplicated to the neighboring edges across those boundaries. That is, if we have a PEC boundary segment passing through a grid node (up to some tolerance) , we consider that this segment crosses both edges around this grid node. This seems to nicely improve the quality of resolving gaps and strips. Here's the circularly polarized antenna with base resolutions 6 (needed 2 iterations), 10, 15, and 20
<img width="376" alt="Screenshot_20250506_220238" src="https://github.com/user-attachments/assets/775a9b72-173d-415c-b222-6ea2ae5c6204" />
<img width="376" alt="antenna10" src="https://github.com/user-attachments/assets/390230c4-6c88-4c42-a14c-e9a62fbc3e72" />
<img width="372" alt="antenna15" src="https://github.com/user-attachments/assets/5bb5f090-f889-4517-8968-c7cf220ac197" />
<img width="371" alt="antenna20" src="https://github.com/user-attachments/assets/ec844834-bd93-46d1-92c2-e85bf4aa32b6" />

- Additionally, now I also explicitly filter out intersections that happen due to very small features slightly protruding into cells. This prevents `dl_min` from becoming unreasonably small. If we do want to resolve such feature in certain cases, it can be done by turning on `convex_resolution` in `CornerFinderSpec`
- Changed the strategy of how snapping points are generated: previously, I would create only 1 snapping line per problematic grid spacing. Now, for a given column (or row) of cells, I look for the one with max number of found intersections and create (num_intersections - 1) snapping points to resolve all gaps/stripes with one go. This seems to reduce the total number of iteration to converge the grid.
<img width="946" alt="Screenshot_20250506_213243" src="https://github.com/user-attachments/assets/6fb2f95f-a2de-48b3-8358-82db19ce3b04" />

weiliangjin2021 commented on tidy3d/components/simulation.py:
```diff
@@ -1142,10 +1147,40 @@ def grid(self) -> Grid:
             internal_override_structures=self.internal_override_structures,
         )
 
+        # This would AutoGrid the in-plane directions of the 2D materials
```
some typos?

dbochkov-flexcompute replied:
hmm, not sure what this means exactly, copied it from the old function

weiliangjin2021 commented on tidy3d/components/grid/grid_spec.py:
```diff
@@ -1037,6 +1043,19 @@ class LayerRefinementSpec(Box):
+    dl_min_from_gap_width: bool = pd.Field(
+        True,
+        title="Set ``dl_min`` from Gap Width",
+        description="Take into account autodetected minimal PEC gap width when determining ``dl_min``.",
```
add description that it only takes effect if dl_min of AutoGrid is not set

weiliangjin2021 commented on tidy3d/components/simulation.py:
```diff
@@ -1118,7 +1123,7 @@ def set_plot_params(boundary_edge, lim, side, thickness):
 
     @cached_property
-    def grid(self) -> Grid:
+    def _grid_and_snapping_lines(self) -> Grid:
```
return type to be updated

weiliangjin2021 reviewed (approved):
Code looks good! Just one part not fully understand: how re-entering subcell features are filtered. Could you provide some visual explanations?

dmarek-flex commented on tidy3d/components/grid/grid_spec.py:
```diff
@@ -1510,6 +1553,346 @@ def _override_structures_along_axis(
+                    h_cells_dx.append(data[3])
+
+                # in case the polygon has holes
+                for poly_inner in poly.interiors:
```
Add a unit test to test this case

dmarek-flex reviewed (approved):
Everything looks pretty good! Just one improvement for code coverage, and friendly reminder for Changelog.

Also I tried another notebook with the gap meshing. [WidebandBeamSteerableReflectarrayWithPRUC.zip](https://github.com/user-attachments/files/20172145/WidebandBeamSteerableReflectarrayWithPRUC.zip)

I assumed it would place snapping lines in the gap between simulation boundary and the arrows. Instead it gives:

![image](https://github.com/user-attachments/assets/9e18f659-c5b1-408c-afd6-9983d060c1ff)

And when I use many more iterations:

![image](https://github.com/user-attachments/assets/0472fce7-4f10-4d4a-9a74-a4d8368a637c)

Is this expected?

dbochkov-flexcompute commented:
> Is this expected?

yeah, currently it ignores boundary conditions. I should add that. In this example, it's periodic I assume? I think need to add for PEC and PMC boundary conditions as well

dmarek-flex commented:
> > Is this expected?
> 
> yeah, currently it ignores boundary conditions. I should add that. In this example, it's periodic I assume? I think need to add for PEC and PMC boundary conditions as well

yea periodic

dbochkov-flexcompute commented:
> > > Is this expected?
> > 
> > 
> > yeah, currently it ignores boundary conditions. I should add that. In this example, it's periodic I assume? I think need to add for PEC and PMC boundary conditions as well
> 
> yea periodic

Actually, just realized, that even with proper boundary treatment there won't be any additional grid lines in this example. Because  the edges across periodic sides are already separated by a grid line (=simulation domain edge)

dbochkov-flexcompute commented:
> Code looks good! Just one part not fully understand: how re-entering subcell features are filtered. Could you provide some visual explanations?

@weiliangjin2021 Damian's case is a good example of that: note those sharp corners that cross the same grid segment before crossing anything else, this is what I call a small re-entry feature. 

<img width="400" alt="Screenshot_20250512_161047" src="https://github.com/user-attachments/assets/5997c477-3339-4c57-abfd-77c49a016fa8" />

When we look for intersections between PEC polygons and grid lines, we go around the polygon's perimeter one segment after another. So, when such a small feature intersects the same edge twice, the array that stores indices of intersected cells will have two consecutive elements with the same value. Such elements are removed from array of intesections.

Btw, in Damian's comment above these small features were still resolved after two iterations :sweat_smile:, but I already found a bag, and now grid stay the same after the first iteration

dbochkov-flexcompute committed:
automatic iterative gap meshing

dbochkov-flexcompute commented:
Should be ready to go. This is how it works now for boundaries:
- refine is PEC or PMC
![Screenshot_20250512_230429](https://github.com/user-attachments/assets/451f742c-0df5-43b9-b489-c34f9307ee0c)

- no refine if periodic and separated
<img width="400" alt="Screenshot_20250512_161047" src="https://github.com/user-attachments/assets/df187c17-3149-4410-b94f-9d012955092a" />

- refine if periodic and no gap on one side
![Screenshot_20250512_230414](https://github.com/user-attachments/assets/98c498a2-b553-4086-8b18-086adfbc6308)

dmarek-flex commented:
Great I think its ready, thanks for handling BCs!
